### PR TITLE
Add AWS and Terraform Tutorial

### PR DESCRIPTION
**Pull request by ajeetraina**

### CODE REVIEW

File: docs/create/aws/terraform/index-terraform.mdx [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+sidebar_label: Deploy and Manage Redis Database on AWS using Terraform
+slug: /create/aws/terraform
+authors: [ajeet,rahul]
+---
+
+![terraform](terraform_arch.png)
+
+
+Development teams today are embracing more and more DevOps principles, such as continuous integration and continuous delivery (CI/CD). Therefore, the need to manage infrastructure-as-code (IaC) has become an essential capability for any cloud service.  IaC tools allow you to manage infrastructure with configuration files rather than through a graphical user interface. IaC allows you to build, change, and manage your infrastructure in a safe, consistent, and repeatable way by defining resource configurations that you can version, reuse, and share. 
+A leading tool in the IaC is HashiCorp Terraform, which supports  the major cloud providers and services with its providers and modules cloud infrastructure automation ecosystem for provisioning, compliance, and management of any cloud, infrastructure, and service.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
A leading tool in the IaC is HashiCorp Terraform, which supports the major cloud providers and services with its providers and modules cloud infrastructure automation ecosystem for provisioning, compliance, and management of any cloud, infrastructure, and service.
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+Terraform is an open source IaC software tool that provides a consistent CLI workflow to manage hundreds of cloud services. Terraform codifies cloud APIs into declarative configuration files, which can then be shared amongst team members, treated as code, edited, reviewed, and versioned. It enables you to safely and predictably create, change, and improve infrastructure. 
+
+### Capabilities of Terraform
+
+- Terraform is not just a configuration management tool. It also focuses on the higher-level abstraction of the data center and associated services, while allowing you to use configuration management tools on individual systems.
+- It supports multiple cloud providers, such as AWS, GCP, Azure, DigitalOcean, etc.
+- It provides a single unified syntax, instead of requiring operators to use independent and non-interoperable tools for each platform and service.
+- Manages both existing service providers and custom in-house solutions.
+- Terraform is easily portable to any other provider.
+- Provides immutable infrastructure where configuration changes smoothly
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
- Provides immutable infrastructure where configuration changes smoothly.
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+Before we jump into the implementation, let us take a moment to better understand the Terraform configuration. A Terraform configuration is a complete document in the Terraform language that tells Terraform how to manage a given collection of infrastructure. A configuration can consist of multiple files and directories. Terraform is broken down into three main components:
+
+- Providers
+- Data sources
+- Resources
+
+
+### Providers
+
+Provider is the first resource that will need to be defined in any project under the Terraform configuration file. The provider gives you access to the API you will be interacting with to create resources. Once the provider has been configured and authenticated, a vast amount of resources are now available to be created. Terraform has more than 100+ cloud providers it serves.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
A provider is the first resource that will need to be defined in any project under the Terraform configuration file. The provider gives you access to the API you will be interacting with to create resources. Once the provider has been configured and authenticated, a vast amount of resources are now available to be created. Terraform has more than 100+ cloud providers it serves. 
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+ count = 2
+ length = 20
+ upper = true
+ lower = true
+ number = true
+}
+```
+
+
+Resource {} block has two strings before the block: resource types and resource names. The prefix of the type maps to the name of the provider. For example, the resource type “random_password” and the resource name “passwords” form a unique identifier of the resource. Terraform uses this ID to identify the resource.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
The resource {} block has two strings before the block: resource types and resource names. The prefix of the type maps to the name of the provider. For example, the resource type “random_password” and the resource name “passwords” form a unique identifier of the resource. Terraform uses this ID to identify the resource.
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+
+
+
+A data block requests that Terraform read from a given data source ("rediscloud_payment_method") and export the result under the given local name ("card"). The name is used to refer to this resource from elsewhere in the same Terraform module, but has no significance outside of the scope of a module.
+Within the block body (between { and }) are query constraints defined by the data source. Most arguments in this section depend on the data source, and indeed in this example card_type and last_four_numbers are all arguments defined specifically for the rediscloud_payment_method data source.
+
+
+### Configure Redis Enterprise Cloud programmatic access
+
+In order to set up authentication with the Redis Enterprise Cloud provider, a programmatic API key must be generated for Redis Enterprise Cloud. The Redis Enterprise Cloud documentation contains the most up-to-date instructions for creating and managing your key(s) and IP access.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx:122]: Might be an idea to link to the relevant Redis Cloud docs here?

File: docs/create/aws/terraform/index-terraform.mdx [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+
+Please note that Flexible and Annual Redis Enterprise Cloud subscriptions can leverage a RESTful API that permits operations against a variety of resources, including servers, services, and related infrastructure. The REST API is not supported for Fixed or Free subscriptions.
+
+```
+ provider "rediscloud" { } # Example resource configuration 
+ resource "rediscloud_subscription" "example" { # ... }
+```
+
+
+### Prerequisite:
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
### Prerequisites:
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+
+- Install Terraform on your MacOS.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
- Install Terraform on MacOS.
```

Comment (simonprickett) [reply docs/create/aws/terraform/index-terraform.mdx]: Might be worth generalizing this as Terraform doesn't only work on a Mac?

File: docs/create/aws/terraform/index-terraform.mdx [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+- Create a free Redis Enterprise Cloud account.
+- Create your first subscription.
+- Enable API
+
+
+### Step 1: Install Terraform on your MacOS
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
### Step 1: Install Terraform on MacOS
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+
+Use Homebrew to install terraform on MacOS as shown below:
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
Use Homebrew to install Terraform on MacOS as shown below:
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 10]
```diff
@@ -0,0 +1,597 @@
...
+
+```
+ brew install terraform
+```
+
+### Step 2: Sign up for a free Red Enterprise Cloud account
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
### Step 2: Sign up for a free Redis Enterprise Cloud account
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 11]
```diff
@@ -0,0 +1,597 @@
...
+  password = random_password.passwords[1].result
+}
+}
+```
+
+
+
+### Step 7: Create an execution plan
+
+The terraform plan command creates an execution plan, which lets you preview the changes that Terraform plans to make to your infrastructure. By default, when Terraform creates a plan, it reads the current state of any already existing remote objects to make sure that Terraform state is up to date. It then compares the current configuration to the prior state and then proposes a set of change actions that should make the remote object match the configuration.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
The Terraform plan command creates an execution plan, which lets you preview the changes that Terraform plans to make to your infrastructure. By default, when Terraform creates a plan, it reads the current state of any already existing remote objects to make sure that Terraform state is up to date. It then compares the current configuration to the prior state and then proposes a set of change actions that should make the remote object match the configuration.
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
+
+Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run "terraform apply" now.
+```
+
+
+
+### Step 8: Execute the action
+
+The terraform apply command executes the actions proposed in a Terraform plan.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
The Terraform apply command executes the actions proposed in a Terraform plan.
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 13]
```diff
@@ -0,0 +1,597 @@
...
+
+### Step 9: Verify the database
+
+You can now verify the new database created under Subscription named “db-json.”
+
+
+
+
+
+Deploy Redis Database with RedisJSON modules on AWS using Terraform
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
Deploy a Redis Database with RedisJSON modules on AWS using Terraform:
```

File: docs/create/aws/terraform/index-terraform.mdx [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+  throughput_measurement_value = 10000
+  password = random_password.passwords[1].result
+}
+}
+```
+
+
+### Step 10: Cleanup 
+
+The terraform destroy command is a convenient way to destroy all remote objects managed by a particular Terraform configuration. While you will typically not want to destroy long-lived objects in a production environment, Terraform is sometimes used to manage ephemeral infrastructure for development purposes, in which case you can use terraform destroy’ to conveniently clean up all of those temporary objects once you are finished with your work.
```
Comment (simonprickett) [inline docs/create/aws/terraform/index-terraform.mdx]:
```suggestion
The Terraform destroy command is a convenient way to destroy all remote objects managed by a particular Terraform configuration. While you will typically not want to destroy long-lived objects in a production environment, Terraform is sometimes used to manage ephemeral infrastructure for development purposes, in which case you can use terraform destroy’ to conveniently clean up all of those temporary objects once you are finished with your work.
```

**Review by simonprickett — APPROVED** [review]
Approved with comments - please check / resolve before merging, thanks!

File: docs/create/aws/terraform/index-terraform.mdx (hunk 5 above)
Comment (ajeetraina) [reply docs/create/aws/terraform/index-terraform.mdx:122]: Let me try to see if I can add link to Redis Enterprise cloud tutorial here.

Comment (ajeetraina) [reply docs/create/aws/terraform/index-terraform.mdx:122]: I have put the reference to our Redis Enterprise Cloud page. Have a look. @simonprickett

**Pull request merged**